lib/ansible/modules/cloud/amazon/aws_sgw_facts.py

```diff
@@ -0,0 +1,358 @@
+    def _read_gateway_tape_response(tapes, aws_response):
+        for tape in aws_response["TapeInfos"]:
+            tape_obj = camel_dict_to_snake_dict(tape)
+            if "gateway_arn" in tape_obj:
+                del tape_obj["gateway_arn"]
+            tapes.append(tape_obj)
+
+        return aws_response["Marker"] if "Marker" in aws_response else None
+
+    """
+    List VTL & VTS attached to AWS storage gateway in VTL mode
+    """
+    def list_gateway_vtl(self, gateway):
+        response = self.connection.list_tapes(
+            Limit=100
+        )
+
+        gateway["tapes"] = []
+        marker = self._read_gateway_tape_response(gateway["tapes"], response)
+
+        while marker is not None:
+            response = self.connection.list_tapes(
+                Marker=marker,
+                Limit=100
+            )
+
+            marker = self._read_gateway_tape_response(gateway["tapes"], response)
+
+    """
+    Read tape objects from AWS API response.
+    Drop the gateway_arn & gateway_id attribute from response, as it will be duplicate 
+    with parent object.
+    """
+    @staticmethod
+    def _read_gateway_volumes(volumes, aws_response):
+        for volume in aws_response["VolumeInfos"]:
+            volume_obj = camel_dict_to_snake_dict(volume)
+            if "gateway_arn" in volume_obj:
+                del volume_obj["gateway_arn"]
+            if "gateway_id" in volume_obj:
+                del volume_obj["gateway_id"]
+
+            volume.append(volume_obj)
+        return aws_response["Marker"] if "Marker" in aws_response else None
+
+    """
+    List volumes attached to AWS storage gateway in CACHED or STORAGE mode
+    """
+    def list_gateway_volumes(self, gateway):
+        response = self.connection.list_volumes(
+            GatewayARN=gateway["gateway_arn"],
+            Limit=100
+        )
+
+        gateway["volumes"] = []
+        marker = self._read_gateway_volumes(gateway["volumes"], response)
+
+        while marker is not None:
+            response = self.connection.list_volumes(
+                GatewayARN=gateway["gateway_arn"],
+                Marker=marker,
+                Limit=100
+            )
+
+            marker = self._read_gateway_volumes(gateway["volumes"], response)
+
+
+def main():
+    argument_spec = ec2_argument_spec()
+
+    module = AnsibleAWSModule(argument_spec=argument_spec)
+
+    if not HAS_BOTO3:
+        module.fail_json(msg='boto3 required for this module')
+
+    region, ec2_url, aws_connect_params = get_aws_connection_info(module, boto3=True)
+
+    connection = boto3_conn(
+        module,
+        conn_type='client',
+        resource='storagegateway',
+        region=region,
+        endpoint=ec2_url,
+        **aws_connect_params
+    )
+
+    if connection is None:  # this should never happen
+        module.fail_json(msg='Unknown error, failed to create glacier connection, no information from boto.')
+
+    SGWFactsManager(connection, module).fetch()
```

Comment:
I think that having this class complicates this module, rather than helps it.

Reply:
i tend to prefer using objects :)